NonTerminals.cs

```diff
@@ -0,0 +1,162 @@
+     * method IsOptionDependent(p)
+     * --------------------------------------------------------------------------
+     * Returns true if p is dependent on any compiler option, else false.
+     * ----------------------------------------------------------------------- */ //SAM CHANGE
+
+    bool IsOptionDependent(Production p) {
+        return p >= firstOptionDependent;
+    } /* end IsOptionDependent */
+
+
+    /* --------------------------------------------------------------------------
+     * method IsConstParamDependent(p)
+     * --------------------------------------------------------------------------
+     * Returns true if p is dependent on CONST parameter option, else false.
+     * ----------------------------------------------------------------------- */ //SAM CHANGE
+
+    bool IsConstParamDependent(Production p)
+    {
+        return p >= firstConstParamDependent && p <= lastConstParamDependent;
+    } /* end IsConstParamDependent */
+
+
+    /* --------------------------------------------------------------------------
+     * method IsVariantRecordDependent(p)
+     * --------------------------------------------------------------------------
+     * Returns true if p is dependent on variant record type option, else false.
+     * ----------------------------------------------------------------------- */ //SAM CHANGE
+
+    bool IsVariantRecordDependent(Production p) {
+        return p >= firstNoVariantRecDependent && p <= lastNoVariantRecDependent;
+    }
+
+    /* --------------------------------------------------------------------------
+     * method FIRST(p)
+     * --------------------------------------------------------------------------
+     * Returns a tokenset with the FIRST set of production p.
+     * ----------------------------------------------------------------------- */
+
+    TokenSet FIRST(Production p) {
+        TokenSet tokenset = null;
+
+        if (IsConstParamDependent(p))
+        {
+            p += alternateSetOffset;
+        } /* end if */
+        if (IsVariantRecordDependent(p) && CompilerOptions.VariantRecords())
+        {
+            p += alternateSetOffset;
+        } /* end if */
+
+        return tokenset;
+    } /* end FIRST */
+
+
+    /* --------------------------------------------------------------------------
+     * method FOLLOW(p)
+     * --------------------------------------------------------------------------
+     * Returns a tokenset with the FOLLOW set of production p.
+     * ----------------------------------------------------------------------- */
+
+    TokenSet FOLLOW(Production p) {
+        TokenSet tokenset = null;
+
+        if (IsConstParamDependent(p)) {
+            p += alternateSetOffset;
+        } /* end if */
+        if (IsVariantRecordDependent(p) && !CompilerOptions.VariantRecords()) {
+            p += alternateSetOffset;
+        } /* end if */
+
+        return tokenset;
+    } /* end FOLLOW */

```

Comment:
This too, always returns null.

We need initialisers for all the FOLLOW sets, in the same manner as for the FIRST sets.

see
https://bitbucket.org/trijezdci/m2c-rework/src/tip/m2-production.c#m2-production.c-139
https://bitbucket.org/trijezdci/m2c-rework/src/tip/m2-production.c#m2-production.c-325

and accordingly ...

```csharp
index = p + alternateSetOffset;
...
tokenset = followSet[index];
...
```